### PR TITLE
Implement `getText` for extension API

### PR DESCRIPTION
**Pull request by tjkandala**
I wasn't able to find a method like [`getText` from VS Code's extension API](https://code.visualstudio.com/api/references/vscode-api#TextDocument), so I implemented this POC.

Example usage:
```ts
function provideHover(document, position) {
    const word = document.getText(document.getWordRangeAtPosition(position))
    // Elided
}
```

#### Motivation

I went through the extension registry and found only two non-"hello world" extensions that registered hover providers. Both were authored by Sourcegraph contributors and implemented their own version of `getText`, but without the utilities available to `TextDocument` methods ([token-highlights](https://sourcegraph.com/github.com/sourcegraph/sourcegraph-extension-samples/-/blob/token-highlights/src/extension.ts#L3), [string-references](https://sourcegraph.com/github.com/lguychard/sourcegraph-string-references/-/blob/src/util.ts#L35)).

Implementing more helpers would make authoring extensions more accessible and reduce extension bundle sizes due to code reuse.

### CODE REVIEW

File: client/packages/sourcegraph-extension-api/src/sourcegraph.d.ts [hunk 1]
```diff
@@ -716,7 +724,7 @@ declare module 'sourcegraph' {
          * Display the results of the location provider (with the given ID) in this panel below the
          * {@link PanelView#contents}.
          *
-         * @internal Experimental. Subject to change or removal without notice.
```
Comment (felixfbecker) [inline client/packages/sourcegraph-extension-api/src/sourcegraph.d.ts:719]: **issue**: we need to keep this note

File: client/packages/sourcegraph-extension-api/src/sourcegraph.d.ts [hunk 2]
```diff
@@ -373,6 +373,14 @@ declare module 'sourcegraph' {
          * @returns A range spanning a word, or `undefined`.
          */
         getWordRangeAtPosition(position: Position): Range | undefined
+
+        /**
+         * Get the substring of text content from the provided range.
+         *
+         * @param range A range
+         * @returns The text inside the provided range, or `undefined`
```
Comment (felixfbecker) [inline client/packages/sourcegraph-extension-api/src/sourcegraph.d.ts:381]: **nitpick**: this doesn't really tell more than the type signature. I'd document in which _case_ it would return `undefined`.

File: client/shared/src/api/extension/api/textDocument.ts [hunk 3]
```diff
@@ -105,6 +105,32 @@ export class ExtensionDocument implements sourcegraph.TextDocument {
         return undefined
     }
 
+    public getText(range?: sourcegraph.Range): string | undefined {
+        this.throwIfNoModelText()
+        range = range ? this.validateRange(range) : undefined
+        if (!range) {
+            return undefined
```
Comment (felixfbecker) [inline client/shared/src/api/extension/api/textDocument.ts:112]: **suggestion**: if no range is provided, I think I would expect to get the whole text? What's the use case for making the parameter optional at all?

File: client/shared/src/api/extension/api/textDocument.test.ts [hunk 4]
```diff
@@ -32,6 +32,23 @@ describe('ExtensionDocument', () => {
...
+        const document = textDocument('aa bb\nbbb ccccc')
+        expect(document.getText(new Range(new Position(0, 3), new Position(1, 3)))).toEqual('bbbbb')
+    })
+
+    test('getText (several lines)', () => {
+        const document = textDocument('this text\ndocument spans\nmultiple \nlines ')
+        expect(document.getText(new Range(new Position(0, 5), new Position(3, 2)))).toEqual(
+            'textdocument spansmultiple li'
+        )
+    })
```
Comment (felixfbecker) [inline client/shared/src/api/extension/api/textDocument.test.ts:51]: **issue**: I think we'd need some test cases here for edge cases - in particular, if the range is out of bounds for the file content (either on the line or the column).

**Review by felixfbecker — COMMENTED** [review]
This is awesome! Way to go!

File: client/shared/src/api/extension/api/textDocument.test.ts (hunk 4 above)
Comment (tjkandala) [reply client/shared/src/api/extension/api/textDocument.test.ts:51]: I had left this test out since `validateRange` and `validatePosition` ensure that the range cannot be out of bounds, but I'll add a test to prevent regressions

File: client/shared/src/api/extension/api/textDocument.ts (hunk 3 above)
Comment (tjkandala) [reply client/shared/src/api/extension/api/textDocument.ts:112]: Range is optional for easy composition with `getWordRangeAtPosition`, which can return undefined. I agree that it should return the whole text in this case, though.

File: client/shared/src/api/extension/api/textDocument.test.ts (hunk 4 above)
Comment (felixfbecker) [reply client/shared/src/api/extension/api/textDocument.test.ts:51]: Oh, in that case you're right, we shouldn't duplicate test cases. I comment here would be enough pointing that out.

File: client/packages/sourcegraph-extension-api/src/sourcegraph.d.ts (hunk 1 above)
Comment (tjkandala) [reply client/packages/sourcegraph-extension-api/src/sourcegraph.d.ts:719]: This was automatically removed (`@internal should be empty eslint(jsdoc/empty-tags)`). I'll move this to the body of the comment.

File: client/shared/src/api/extension/api/textDocument.ts (hunk 3 above)
Comment (tjkandala) [reply client/shared/src/api/extension/api/textDocument.ts]: Returns the whole text now, but text can still be undefined, so the type signature hasn't changed.

**Review by felixfbecker — APPROVED** [review]
